### DISCUSSION TITLE
Why I cannot inherit a usercontrol from templatedcontrol?

### DISCUSSION THREAD

**Discussion by kyuranger** [AvaloniaUI/Avalonia#10732] [Q&A]
Here is my code of TemplatedControl:
XAML:
```
<Styles xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:controls="using:AvaloniaApplication1">
  <Design.PreviewWith>
    <controls:TemplatedControl1 />
  </Design.PreviewWith>

  <Style Selector="controls|TemplatedControl1">
    <!-- Set Defaults -->
    <Setter Property="Template">
      <ControlTemplate>
		  <Border Height="100" Width="100" Background="Red">
		  <ContentPresenter Content="{TemplateBinding Content}"></ContentPresenter>
		  </Border>
      </ControlTemplate>
    </Setter>
  </Style>
</Styles>

```

Code-behind:
```
using Avalonia;
using Avalonia.Controls.Primitives;
using Avalonia.Metadata;

namespace AvaloniaApplication1;

public class TemplatedControl1 : TemplatedControl
{
    public static readonly StyledProperty<object> ContentProperty =
   AvaloniaProperty.Register<TemplatedControl, object>(nameof(Content));
    
    public object Content
    {
        get { return GetValue(ContentProperty); }
        set { SetValue(ContentProperty, value); }
    }
}
```

And here is code of the usercontrol:
XAML:
```
<local:TemplatedControl1 xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"			 
             mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
			 xmlns:local="using:AvaloniaApplication1"
             x:Class="AvaloniaApplication1.UserControl1">
	<TextBlock>123</TextBlock>
</local:TemplatedControl1>

```

Code-behind:
```
using Avalonia;
using Avalonia.Controls;
using Avalonia.Markup.Xaml;

namespace AvaloniaApplication1;

public partial class UserControl1 : TemplatedControl1
{
    public UserControl1()
    {
        InitializeComponent();
    }
}
```

Then when I compie the project, it reports this error:

> Severity	Code	Description	Project	File	Line	Suppression State
Error	AVLN:0004	Internal compiler error while transforming node XamlX.Ast.XamlAstObjectNode:
System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index
   at System.ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument argument, ExceptionResource resource)
   at XamlX.Transform.Transformers.ResolveContentPropertyTransformer.Transform(AstTransformationContext context, IXamlAstNode node)
   at XamlX.Transform.AstTransformationContext.Visitor.Visit(IXamlAstNode node) Line 1, position 2.	AvaloniaApplication1	E:\AvaloniaApplication1\UserControl1.axaml	1	

What's wrong with my code?

**Comment by timunie** [comment]
Templated control has no Content property. I think you need to annotate your Content-property with the [Content] attribute for the XAML-Compiler to understand. 

See: https://github.com/AvaloniaUI/Avalonia/blob/a613f77d0b4ad6fa409e1409ee151bcea4978082/src/Avalonia.Controls/ContentControl.cs#L48-L57

Alternatively (and in this case my favorite solution) inherit from `ContentControl` instead, as it already implements everything needed. It inherits `TemplatedControl` by it's own, so should be what you want imo.

**Reply by kyuranger** [reply]
> Templated control has no Content property. I think you need to annotate your Content-property with the [Content] attribute for the XAML-Compiler to understand.
> 
> See:
> 
> https://github.com/AvaloniaUI/Avalonia/blob/a613f77d0b4ad6fa409e1409ee151bcea4978082/src/Avalonia.Controls/ContentControl.cs#L48-L57
> 
> Alternatively (and in this case my favorite solution) inherit from `ContentControl` instead, as it already implements everything needed. It inherits `TemplatedControl` by it's own, so should be what you want imo.

I modify my code like this:
```
using Avalonia;
using Avalonia.Controls;
using Avalonia.Controls.Primitives;
using Avalonia.Controls.Templates;
using Avalonia.Metadata;

namespace AvaloniaApplication1;

public class TemplatedControl1 : TemplatedControl
{
    public static readonly StyledProperty<object?> ContentProperty =
   AvaloniaProperty.Register<TemplatedControl, object?>(nameof(Content));
    [Content]
    [DependsOn(nameof(ContentTemplate))]
    public object? Content
    {
        get { return GetValue(ContentProperty); }
        set { SetValue(ContentProperty, value); }
    }

    public static readonly StyledProperty<IDataTemplate?> ContentTemplateProperty =
           AvaloniaProperty.Register<ContentControl, IDataTemplate?>(nameof(ContentTemplate));

    public IDataTemplate? ContentTemplate
    {
        get { return GetValue(ContentTemplateProperty); }
        set { SetValue(ContentTemplateProperty, value); }
    }
}
```

It compiles without any errors. However, it runs but the usercontrol does not render any.

I have ensured add styleinclude in app.axmal and the templatecontrol which usercontrol inherits from works well also.

**Reply by timunie** [reply]
I still don't get the use case tbh. `UserControl` is a bit special in handling. I wouldn't reinvent the wheel if I were you. Custom controls can be useful for sure, but normally you just do this: 

```xml
<StackPanel>
   <local:TemplatedControl1>
      <MyContent />
   </local:TemplatedControl1>
   <local:TemplatedControl1 Content="Hello World" />
</StackPanel>
```

but if you really need to do what's written above, I see two options: 

1. Choose the `:is()`-Selector: 
```xml
<Style Selector=":is(local|TemplatedControl1)">
    <Setter Property="Template">
        <ControlTemplate>
            <ContentPresenter Name="PART_ContentPresenter" Content="{TemplateBinding Content}" />
        </ControlTemplate>
    </Setter>
</Style>
```
2. Implement `IStyleable` and override the style key

See: https://docs.avaloniaui.net/docs/styling/selectors#is

**Reply by kyuranger** [reply]
> I still don't get the use case tbh. `UserControl` is a bit special in handling. I wouldn't reinvent the wheel if I were you. Custom controls can be useful for sure, but normally you just do this:

Currently I am migrating an existing WPF personal project to Avalonia. What I want to do is just inherit XAML(including controltemplate) and dependency property to it. 

I need to do this because there is a base control A, B\C\D\E control is inherited from A with different content. B\C\D\E will reuse many times that I have to wrap it into an independent control. In WPF I create A with custom control and B\C\D\E inherit from A with different content.

In AVALONIA there is no custom control any more that I create A with templated control and create B\C\D\E by inherit from A again.

**Reply by kyuranger** [reply]
> I still don't get the use case tbh. `UserControl` is a bit special in handling. I wouldn't reinvent the wheel if I were you. Custom controls can be useful for sure, but normally you just do this:
> 
> ```
> <StackPanel>
>    <local:TemplatedControl1>
>       <MyContent />
>    </local:TemplatedControl1>
>    <local:TemplatedControl1 Content="Hello World" />
> </StackPanel>
> ```
> 
> but if you really need to do what's written above, I see two options:
> 
> 1. Choose the `:is()`-Selector:
> 
> ```
> <Style Selector=":is(local|TemplatedControl1)">
>     <Setter Property="Template">
>         <ControlTemplate>
>             <ContentPresenter Name="PART_ContentPresenter" Content="{TemplateBinding Content}" />
>         </ControlTemplate>
>     </Setter>
> </Style>
> ```
> 
> 2. Implement `IStyleable` and override the style key
> 
> See: https://docs.avaloniaui.net/docs/styling/selectors#is

Base on your way, currently I achieve it by this:
The base control:
```
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
			 xmlns:local="using:AvaloniaApplication1"
             x:Class="AvaloniaApplication1.UserControl1">
	<UserControl.Styles>
		<Style Selector=":is(local|UserControl1)">
			<Setter Property="Template">
				<ControlTemplate>
					<Border Background="Red">
						<ContentPresenter Name="PART_ContentPresenter" Content="{TemplateBinding Content}" />
					</Border>
				</ControlTemplate>
			</Setter>
		</Style>
	</UserControl.Styles>
</UserControl>
```

And the inherit Control:
```
<local:UserControl1 xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
			 xmlns:local="using:AvaloniaApplication1"
             x:Class="AvaloniaApplication1.UserControl2">
  Welcome to Avalonia!
</local:UserControl1>
```
```
using Avalonia;
using Avalonia.Controls;
using Avalonia.Markup.Xaml;

namespace AvaloniaApplication1;

public partial class UserControl2 : UserControl1
{
    public UserControl2()
    {
        InitializeComponent();
    }
}
```

It works when usecontrol inherit from usercontrol but cannot when usercontrol inherit from templatecontrol.

I am not sure whether it is bug of Avalonia while WPF can do this.  Or templatecontrol differs from customcontrol much.

**Reply by YunyiZhang** [reply]
If you need to add custom content to your own control, you have to inherit from a ContentControl, while TemplatedControl is not.
